src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2356,57 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static void ucp_worker_keepalive_deactivate(ucp_worker_h worker)
+{
+    uct_worker_progress_unregister_safe(worker->uct,
+                                        &worker->keepalive.cb_id);
+}
+
+static unsigned ucp_worker_keepalive_cb(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    double now          = ucs_get_accurate_time();
+    unsigned cnt;
+    ucp_ep_ext_gen_t *ep_ext;
+    ucp_ep_h ep;
+    ucs_list_link_t *iter_begin;
+
+    if ((now - worker->keepalive.last_shot) <
+        worker->context->config.ext.keepalive_timeout) {
+        return 0;
+    }
+
+    worker->keepalive.last_shot = now;
+    cnt                         = 0;
+    iter_begin                  = worker->keepalive.iter;
+    ucs_list_for_each(ep_ext, iter_begin, ep_list) {
+        worker->keepalive.iter = &ep_ext->ep_list;
+        if (worker->keepalive.iter == &worker->all_eps) {
+            continue;
+        }
+        ep   = ucp_ep_from_ext_gen(ep_ext);
+        cnt += !!ucp_ep_check(ep);
+        if (cnt >= worker->context->config.ext.keepalive_num_eps) {
+            break;
+        }
+    }
+
+    return cnt;
+}
+
+void ucp_worker_keepalive_activate(ucp_worker_h worker)
+{
+    if ((worker->context->config.ext.keepalive_timeout == 0.0) ||
+        (worker->context->config.ext.keepalive_num_eps == 0)) {
+        return;
+    }
+
+    uct_worker_progress_register_safe(worker->uct, ucp_worker_keepalive_cb,
```

Comment:
when last ep with err handling was removed, need to remove the keepalive as well

Reply:
yep, unsubscribe callback on last EP is removed

Reply:
i think we can avoid the callback at all, if all eps are without err handling

Reply:
not sure I got your comment...
When we are building configuration where keepalive is enabled (ucp_ep has non-zero ep_check mask) we are enabling callback. When callback is called - it checks for existing EPs which may test keepalive, and in case if no such EPs - unsubscribe from progress. On EP remove - we are check for existing keepalive EPs too & make decision about unsubscribe.

Is this logic incorrect?